Mathlib/Data/Matrix/Basic.lean

```diff
@@ -0,0 +1,2823 @@
+  ext
+  simp only [diagonal_apply, map_apply]
+  split_ifs <;> simp [h]
+#align matrix.diagonal_map Matrix.diagonal_map
+
+@[simp]
+theorem diagonal_conjTranspose [AddMonoid α] [StarAddMonoid α] (v : n → α) :
+    (diagonal v)ᴴ = diagonal (star v) := by
+  rw [conjTranspose, diagonal_transpose, diagonal_map (star_zero _)]
+  rfl
+#align matrix.diagonal_conj_transpose Matrix.diagonal_conjTranspose
+
+section One
+
+variable [Zero α] [One α]
+
+instance one : One (Matrix n n α) :=
+  ⟨diagonal fun _ => 1⟩
+
+@[simp]
+theorem diagonal_one : (diagonal fun _ => 1 : Matrix n n α) = 1 :=
+  rfl
+#align matrix.diagonal_one Matrix.diagonal_one
+
+theorem one_apply {i j} : (1 : Matrix n n α) i j = if i = j then 1 else 0 :=
+  rfl
+#align matrix.one_apply Matrix.one_apply
+
+@[simp]
+theorem one_apply_eq (i) : (1 : Matrix n n α) i i = 1 :=
+  diagonal_apply_eq _ i
+#align matrix.one_apply_eq Matrix.one_apply_eq
+
+@[simp]
+theorem one_apply_ne {i j} : i ≠ j → (1 : Matrix n n α) i j = 0 :=
+  diagonal_apply_ne _
+#align matrix.one_apply_ne Matrix.one_apply_ne
+
+theorem one_apply_ne' {i j} : j ≠ i → (1 : Matrix n n α) i j = 0 :=
+  diagonal_apply_ne' _
+#align matrix.one_apply_ne' Matrix.one_apply_ne'
+
+@[simp]
+theorem map_one [Zero β] [One β] (f : α → β) (h₀ : f 0 = 0) (h₁ : f 1 = 1) :
+    (1 : Matrix n n α).map f = (1 : Matrix n n β) := by
+  ext
+  simp only [one_apply, map_apply]
+  split_ifs <;> simp [h₀, h₁]
+#align matrix.map_one Matrix.map_one
+
+theorem one_eq_pi_single {i j} : (1 : Matrix n n α) i j = Pi.single (f := fun _ => α) i 1 j := by
+  simp only [one_apply, Pi.single_apply, eq_comm]
+#align matrix.one_eq_pi_single Matrix.one_eq_pi_single
+-- Porting note: (implicit arg) added argument `(f := fun_ => α)` explicitely, why is that needed?
+-- also removed `<;> congr` at the end of the proof.
+
+-- deal with decidable_eq
+end One
+
+section Numeral
+
+set_option linter.deprecated false
+
+@[deprecated, simp]
+theorem bit0_apply [Add α] (M : Matrix m m α) (i : m) (j : m) : (bit0 M) i j = bit0 (M i j) :=
+  rfl
+#align matrix.bit0_apply Matrix.bit0_apply
+
+variable [AddZeroClass α] [One α]
+
+@[deprecated]
+theorem bit1_apply (M : Matrix n n α) (i : n) (j : n) :
+    (bit1 M) i j = if i = j then bit1 (M i j) else bit0 (M i j) := by
+  dsimp [bit1]
+  by_cases h : i = j <;>
+  simp [h]
+#align matrix.bit1_apply Matrix.bit1_apply
+
+@[deprecated, simp]
+theorem bit1_apply_eq (M : Matrix n n α) (i : n) : (bit1 M) i i = bit1 (M i i) := by
+  simp [bit1_apply]
+#align matrix.bit1_apply_eq Matrix.bit1_apply_eq
+
+@[deprecated, simp]
+theorem bit1_apply_ne (M : Matrix n n α) {i j : n} (h : i ≠ j) : (bit1 M) i j = bit0 (M i j) := by
+  simp [bit1_apply, h]
+#align matrix.bit1_apply_ne Matrix.bit1_apply_ne
+
+end Numeral
+
+end Diagonal
+
+section Diag
+
+/-- The diagonal of a square matrix. -/
+-- @[simp] -- Porting note: simpNF does not like this.
+def diag (A : Matrix n n α) (i : n) : α :=
+  A i i
```

Comment:
If this is a `simp`-Lemma then all the following lemmas about `diag` at "not in simp normal form", so I just removed the simp on this definition for now...

Reply:
It looks like we need an additional `diag_apply` lemma that is `simp`